### ISSUE TITLE
Allow users to set blockingTaskExecutor per GrpcService

### ISSUE THREAD

**Issue by vkostyukov** [line/armeria#4799]
Looks like this ability already exist for raw HTTP services, via the [`.route()`](https://armeria.dev/docs/server-basics/#per-service-configuration) API.

It would be nice (and needed for my use-case) to unlock the same functionality for gRPC services. There is already [GrpcServiceBuilder.useBlockingTaskExecutor:  Boolean](https://javadoc.io/doc/com.linecorp.armeria/armeria-javadoc/latest/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.html#useBlockingTaskExecutor(boolean)), what's left is to figure out how to put a user-defined blocking tasks executor into a request context. One (very generic) way to do is it to expose `.setBlockingTaskExecutor` on `ServiceRequestContext`. Another Option is control this functionality via `GrpcServiceBuilder`, perhaps by adding `blockingTaskExecutor(executor)` method there.

I'm not super sure if route-like API could exist for HTTP so I didn't list it as an option. Maybe it could?

**Comment by jrhee17** [comment]
I think the original intention is that `blockingTaskExecutor` set from `ServiceConfigBuilder` can be used for each gRPC service.

Currently, the `ServerBuilder#route` API can be used to declare gRPC services in the following way:

1. Adding each route
```
sb.route()
  .blockingTaskExecutor(Executors.newScheduledThreadPool(3), true)
  .build(grpcService);
```

2. Using `pathPrefix`
```
sb.route()
  .pathPrefix("/api")
  .blockingTaskExecutor(Executors.newScheduledThreadPool(3), true)
  .build(grpcService);
```

Having said this, the behavior isn't well defined/documented. For instance, setting `ServiceConfigBuilder#path` will prevent routes from being added.

Moreover, it isn't possible to:
- Add a prefix to each service
- Transcoding isn't available when gRPC services are defined this way.

**Comment by vkostyukov** [comment]
Thanks @jrhee17!

I played around with this idea: 

```scala
sb.route()
  .blockingTaskExecutor(Executors.newScheduledThreadPool(3), true)
  .build(grpcService);
```

And I think it should actually work as expected (assuming we also set the `useBlockingTaskExecutor` flag on the GrpcServiceBuilder) and register the gRPC routes.

Another thing I was hoping to explore is how involved it would be to use per-method task executor. I think it comes down to the idea of exposing `ServiceRequestContext.setBlockingTaskExecutor` API and allow users to potentially select appropriate task executor within the decorator-based middleware. 

What do you think about this ^ idea, @jrhee17? The new API we'd need to expose seems to be pretty low footprint and is also fitting nicely with the rest of the capabilities exposed via `ServiceRequestContext`.

**Comment by jrhee17** [comment]
> exposing ServiceRequestContext.setBlockingTaskExecutor API and allow users to potentially select appropriate task executor within the decorator-based middleware

I think this is a reasonable API proposal.
`ServiceRequestContext.blockingTaskExecutor` can be populated with the correct executor, and we would just allow users to override the executor.

Do you have any thoughts @trustin @minwoox @ikhoon ?

**Comment by ikhoon** [comment]
👍 It looks a good idea so that users dynamically choose a block task executor based on the current request context.

**Comment by trustin** [comment]
I think it's fine to add `setBlockingTaskExecutor()` as long as we enforce the following restrictions to prevent the situation where `blockingTaskExecutor()` returns two different instances:

- `setBlockingTaskExecutor()` throws an `IllegalStateException` if `blockingTaskExecutor()` or `setBlockingTaskExecutor()` was invoked before.
- Exception: `setBlockingTaskExecutor()` doesn't throw an exception if the caller specify the same instance with the current one.

**Comment by trustin** [comment]
An alternative option is to provide a way to create a service per gRPC method, e.g.

```java
List<GrpcService> services = GrpcService
  .builder()
  ...
  .buildPerMethod();

services.forEach(s => {
  serverBuilder
    .route()
    .blockingTaskExecutor(...)
    .build(s);
});
```

Then we don't have to introduce a setter method to `ServiceRequestContext`.

**Comment by jrhee17** [comment]
> An alternative option is to provide a way to create a service per gRPC method, e.g.

I see, I think this might be a better way to handle this.
I did think the behavior of setting the `blockingTaskExecutor` is slightly dependent on timing, and if multiple threads are involved what happens downstream may be not well defined.

**Comment by vkostyukov** [comment]
I agree. I think it would be nice and helpful if we can return `Map[Route, GrpcService]` from `buildPerMethod()`. Let me think more about this.

**Comment by trustin** [comment]
A map of Route and GrpcService also sounds good to me. GrpcService implements HttpServiceWithRoutes so you can get its routes via `GrpcService.routes()` though.

**Comment by ikhoon** [comment]
Another possible solution is to specify a `BlockingTaskExecutor` factory to `@Blocking` annotation.

```java
class MyBlockingTaskExecutorFactory implements BlockingTaskExecutorFactory {

    BlockingTaskExecutor blockingTaskExecutor(ServiceRequestContext ctx) {
        return myBlockingTaskExecutor;
    }
}

@Blocking(MyBlockingTaskExecutorFactory.class)
myBlockingMethod(...) {

}
```

**Comment by trustin** [comment]
@ikhoon That also will work. It will also work better for annotated services. However, I'm somewhat concerned about such a change (similarly to the above `setBlockingTaskExecutor()` idea) because it breaks a user's assumption that `ctx.serviceConfig() == ctx.blockingTaskExecutor()`.

**Comment by vkostyukov** [comment]
Quick update on my end:

I was able to get it working using the pattern suggested by @trustin. Here is [a gist](https://gist.github.com/vkostyukov/4e6740d6b58385dea0963952aed3ba31) of what I used internally. It's very close, if not identical, to the `buildPerMethod()` idea discussed before. I guess `buildPerMethod()` makes this pattern more approachable as users don't need to manually pair a route with the service.
